infrastructure/code/convert.go

```diff
@@ -559,82 +558,118 @@ func (s *SarifConverter) getMarkers(r codeClientSarif.Result, baseDir types.File
-func createAutofixWorkspaceEdit(absoluteFilePath string, fixedSourceCode string) (edit types.WorkspaceEdit) {
-	fileContent, err := os.ReadFile(absoluteFilePath)
-	if err != nil {
+// CreateWorkspaceEditFromDiff turns the returned fix (in diff format) into a series of TextEdits in a WorkspaceEdit.
+func CreateWorkspaceEditFromDiff(absoluteFilePath string, diff string) (edit types.WorkspaceEdit) {
+	fileContentBytes, err := os.ReadFile(absoluteFilePath)
+	if err != nil || len(fileContentBytes) == 0 {
 		return edit
 	}
-	singleTextEdit := types.TextEdit{
-		Range: types.Range{
-			// TODO(alex.gronskiy): should be changed to an actual hunk-like edit instead of
-			// this "replace-the-whole-file" strategy
-			Start: types.Position{
-				Line:      0,
-				Character: 0},
-			End: types.Position{
-				Line:      math.MaxInt32,
-				Character: 0},
-		},
-		NewText:  fixedSourceCode,
-		FullText: string(fileContent),
+	fileContentLineStrings := strings.Split(string(fileContentBytes), "\n")
+
+	var textEdits []types.TextEdit
+
+	var hunkLine = 0   // Location in the current hunk
+	var hunkOffset = 0 // Whether we need to offset the current hunk based on previous edits.
+
+	// Loop over the diff. Diffs will always use \n instead of \r\n, so no need to sanitize (see getUnifiedDiff).
+	for _, line := range strings.Split(diff, "\n") {
+		// If we are being asked to make changes outside the original file, abort and return an empty edit.
+		if hunkLine-hunkOffset > len(fileContentLineStrings) {
+			return edit
+		}
+
+		if strings.HasPrefix(line, "---") || strings.HasPrefix(line, "+++") {
+			continue // We ignore header lines
+		} else if strings.HasPrefix(line, "@@") {
+			r := regexp.MustCompile(`@@ -(\d+),?(\d*) \+(\d+),?(\d*) @@`)
+			matches := r.FindStringSubmatch(line)
+			if matches == nil {
+				return edit // We have a badly formatted diff, so abort.
+			}
+
+			hunkLine, _ = strconv.Atoi(matches[1]) // Apply the edit from the first line of the original file in the diff
+			hunkLine -= 1                          // TextEdit range is 0-indexed, whereas LSP diff is 1-indexed
+			hunkLine += hunkOffset                 // Account for any previous additions or deletions.
+		} else if strings.HasPrefix(line, "-") {
+			textEdit := buildOneLineTextEdit(hunkLine, hunkLine+1, "")
+			textEdits = append(textEdits, textEdit)
+			hunkOffset -= 1 // We've removed a line, so future hunks will be offset with respect to the diff.
+		} else if strings.HasPrefix(line, "+") {
+			textEdit := buildOneLineTextEdit(hunkLine, hunkLine, strings.TrimPrefix(line, "+")+"\n")
+			textEdits = append(textEdits, textEdit)
+			hunkOffset += 1 // We've added a line, so future hunks will be offset with respect to the diff.
+			hunkLine += 1   // We've added a line, so increment the pointer in our current hunk.
+		} else if strings.HasPrefix(line, " ") {
+			hunkLine += 1 // We skip over unchanged lines, so increment the pointer in our current hunk.
+		}
```

Comment:
I'd like this to use a lib. Dunno if it exists, though.

Reply:
Maybe this provides a solution: https://www.perplexity.ai/search/is-there-a-golang-lib-to-conve-aoVguwy0Q9CmIEBXw4zPNQ